File: src/1415-The-k-th-Lexicographical-String-of-All-Happy-Strings-of-Length-n/1415.cpp
class Solution {
public:
    string getHappyString(int n, int k) {
        int prem = 1 << (n - 1);
        if (k > 3 * prem) return "";
        
        k--;
        string s = string(1, 'a' + k / prem);
        while (prem > 1) {
            k %= prem;
            prem >>= 1;
            s += k / prem == 0 ? 'a' + (s.back() == 'a') : 'b' + (s.back() != 'c');
        }
        return s;
    }
};

File: src/1415-The-k-th-Lexicographical-String-of-All-Happy-Strings-of-Length-n/1415.py
class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        prem = 1 << (n - 1)
        if prem * 3 < k: return ""
        
        k -= 1
        s = chr(97 + k // prem)
        while prem > 1:
            k %= prem
            prem >>= 1
            s += chr(97 + int(s[-1] == 'a')) if k // prem == 0 else chr(98 + int(s[-1] != 'c'))
        return s